drivers/gpio/gpio_emul.c

```diff
@@ -0,0 +1,699 @@
+	const gpio_pin_t num_pins;
+	/** Supported interrupts */
+	const enum gpio_emul_interrupt_cap interrupt_caps;
+};
+
+/**
+ * @brief Emulated GPIO controller data
+ *
+ * This structure contains data structures used by a emulated GPIO
+ * controller.
+ *
+ * If the application wishes to specify a "wiring" for the emulated
+ * GPIO, then a @ref gpio_callback_handler_t should be registered using
+ * @ref gpio_add_callback.
+ *
+ * Changes are to @ref gpio_emul_data and @ref gpio_emul_config are
+ * synchronized using @a mu.
+ */
+struct gpio_emul_data {
+	/** Common @ref gpio_driver_data */
+	struct gpio_driver_data common;
+	/** Pointer to an array of flags is @a num_pins in size */
+	gpio_flags_t *flags;
+	/** Input values for each pin */
+	gpio_port_value_t input_vals;
+	/** Output values for each pin */
+	gpio_port_value_t output_vals;
+	/** Interrupt status for each pin */
+	gpio_port_pins_t interrupts;
+	/** Mutex to synchronize accesses to driver data and config */
+	struct k_mutex mu;
+	/** Singly-linked list of callbacks associated with the controller */
+	sys_slist_t callbacks;
+};
+
+/**
+ * @brief Obtain a mask of pins that match all of the provided @p flags
+ *
+ * Use this function to see which pins match the current GPIO configuration.
+ *
+ * The caller must ensure that @ref gpio_emul_data.mu is locked.
+ *
+ * @param port The emulated GPIO device pointer
+ * @param mask A mask of flags to match
+ * @param flags The flags to match
+ *
+ * @return a mask of the pins with matching @p flags
+ */
+static gpio_port_pins_t
+get_pins_with_flags(const struct device *port, gpio_port_pins_t mask,
+	gpio_flags_t flags)
+{
+	size_t i;
+	gpio_port_pins_t matched = 0;
+	struct gpio_emul_data *drv_data =
+		(struct gpio_emul_data *)port->data;
+	const struct gpio_emul_config *config =
+		(const struct gpio_emul_config *)port->config;
+
+	for (i = 0; i < config->num_pins; ++i) {
+		if ((drv_data->flags[i] & mask) == flags) {
+			matched |= BIT(i);
+		}
+	}
+
+	return matched;
+}
+
+/**
+ * @brief Obtain a mask of pins that are configured as @ref GPIO_INPUT
+ *
+ * The caller must ensure that @ref gpio_emul_data.mu is locked.
+ *
+ * @param port The emulated GPIO device pointer
+ */
+static inline gpio_port_pins_t get_input_pins(const struct device *port)
+{
+	return get_pins_with_flags(port, GPIO_INPUT, GPIO_INPUT);
+}
+
+/**
+ * @brief Obtain a mask of pins that are configured as @ref GPIO_OUTPUT
+ *
+ * The caller must ensure that @ref gpio_emul_data.mu is locked.
+ *
+ * @param port The emulated GPIO device pointer
+ */
+static inline gpio_port_pins_t get_output_pins(const struct device *port)
+{
+	return get_pins_with_flags(port, GPIO_OUTPUT, GPIO_OUTPUT);
+}
+
+/**
+ * Check if @p port has capabilities specified in @p caps
+ *
+ * @param port The emulated GPIO device pointer
+ * @param caps A bitmask of capabilities
+ */
+static inline bool gpio_emul_config_has_caps(const struct device *port,
+		enum gpio_emul_interrupt_cap caps) {
```

Comment:
This is useful, but the `caps` argument needs to be type `int`, because this is invoked with arguments that are the bitwise or of two of the enum values, which is not itself an enum value.

Reply:
True, yea